### ISSUE TITLE
set coherent behavior between atoms and numberboxes

### ISSUE THREAD

**Issue by jmmmp** [pure-data/pure-data#851]
This is an issue that has always been there as far as I can remember, but could be improved: the gui behaviour of atoms and numberboxes isn't the same at least in one detail:

- create an atom
- turn off edit mode
- click in the atom
- type a number and press enter
- type a second number and press enter

This is the normal and expected behaviour. If you do the same with a number box, in the last step the number goes to 0. That is, without a mouse click, any inputted number resets to 0. From my user experience, the same behaviour as an atom is much more logical and faster to work with.

**Comment by umlaeute** [comment]
just to make sure: with "atoms" you mean ordinary numberboxes (*Number*, <kbd>Ctrl</kbd>+<kbd>1</kbd>), and with "numberboxes" you mean iemgui numberboxes (*Number2*, <kbd>Shift</kbd>+<kbd>Ctrl</kbd>+<kbd>n</kbd>)?

**Comment by jmmmp** [comment]
> just to make sure: with "atoms" you mean ordinary numberboxes (*Number*,
> Ctrl+1), and with "numberboxes" you mean iemgui numberboxes (*Number2*,
> Shift+Ctrl+n)?
>
Yes, those were the names in the help files/menus [only, ctrl+3 for number]

**Comment by porres** [comment]
> This is the normal and expected behaviour. If you do the same with a number box, in the last step the number goes to 0. 

sorry, I cannot either understand or reproduce this

**Comment by umlaeute** [comment]
just follow the step-by-step instructions using a gatom "number box" and an iemgui "numberbox2" and note the differences?

**Comment by porres** [comment]
> just follow the step-by-step instructions using a gatom "number box" and an iemgui "numberbox2" and note the differences?

It's not working... I get something else, have you been able to reproduce it?

Here's what I am getting on macOS, for [nbx]. I type a number and press enter, ok, fine, but I can't type a second number!!! I got the idea from the description that you could type a second number and that it "became" a zero, but a more accurate description, from what I see, is "you can't type a second number".

But I see now that if I hit enter a seconds time, I get a "0" out... but it's not like I typed something that "became" 0...

And, as long as we're at it, there are other inconsistencies too...

When I click on an atom box, I can wait forever until I decide to type in a number value, and I can take all eternity if I want until I hit enter, and it'll just wait for me. Now, [nbx] is merciless, it waits for 3 seconds, and if you did not do anything, if you stop to sneeze or yarn or something, it's gone! You gotta reclick and restart the process.

I agree and I don't see why we should have 2 different behaviors and I like the way atom boxes work, where you can keep typing numbers without reclicking, and that it's nice and polite and just waits for you...

**Comment by Spacechild1** [comment]
what also bugs me is that atom boxes don't give you a visual clue when they are ready to enter a new value. numberbox2 will at least temporarily change the front color to red. this is a really issue with patch *users*. When I ask them to click and enter a number, I often get the response "it doesn't work", because they expect some kind of visual clue after they've clicked the atom box.

**Comment by porres** [comment]
how about when you click on it, it gets highlighted as when you select it?

**Comment by porres** [comment]
and here's another thought, we could use for both atom boxes and [nbx] the ability to hit upwards and downwards arrows for controlled single increments upwards and downwards. This would go up or down in single units, or the same as "1 pixel" for [nbx] with a log setting. Alternatively, we can also use "shift" for increments of one hundredth or 0.01 pixels.

**Comment by porres** [comment]
> how about when you click on it, it gets highlighted as when you select it?

hmm, that wouldn't be the same as [nbx], and in fact, what [nbx] does is a bit weird. It has a hardcoded color of red for the number when the box is activated, and this is bad if you have a red background as you can't see the number! (Bad Design). 

We could think of adding a 4th color setting or go to a whole new approach.

Anyway, highlight the value is what would be the same.

BTW, this behavior of changing the object to indicate activity is special to [nbx], no other GUI behaves like that - but I guess it makes sense, since it's the only one that allows you to type in.

**Comment by porres** [comment]
ok, so far we have:
- Remove the activity time of 3 seconds from [nbx], let it keep active until you click out of the object.
- Keep [nbx] active after hitting enter, so it allows you to type another number. 
- Make atom boxes highlight their values when you click on it to indicate activity.

I have another issue, for both. When you click on the object and hit enter, the value becomes zero... but I think it should output the holding value out again... anyway, this seems like a separate issue.

**Comment by porres** [comment]
related to this issue of increasing the consistency, I had requested atom boxes to keep i/olet fucntional in https://github.com/pure-data/pure-data/issues/1005 - and I already have a PR for that in https://github.com/pure-data/pure-data/pull/1004 - maybe I can try and include more commits to attend the issues here.

**Comment by porres** [comment]
> remove unnecessary ability to type "+" in [nbx] 

ACTUALLY........... using "+" is useful and is printed for something like 1e**+**10

so I guess it's the other way around now... an atom number box could also allow you to type in "+"

**Comment by porres** [comment]
> Keep [nbx] active after hitting enter, so it allows you to type another number.

[nbx] actually in fact stays active, it just doesn't show it (as the number goes back from red to the front color). And by remaining active, you can hit enter and it'll respond.

So what we need is to:
- keep it visually active (keep number 'red');
- and also allow you to type in a new value...

**Comment by porres** [comment]
@jmmmp - I think I have already attended your issue, I will now try to work on the thingy @Spacechild1 mentioned (make gatom visually active as well).

Please test!

**Comment by porres** [comment]
>  I will now try to work on the thingy @Spacechild1 mentioned (make gatom visually active as well).

I already give up, even though I really want that too. Maybe @Spacechild1 can do it with much more competence... I see the drawing doesn't really happen in the code of g_text.c, but in a shared function called **glist_retext**, which means we have to create a special function just for the gatom.

**Comment by porres** [comment]
We can also maybe open a new issue to request and discuss this behaviour for atom boxes showing you if it's active, and consider this issue fully closed by https://github.com/pure-data/pure-data/pull/1004

**Comment by porres** [comment]
> how about when you click on it, the value gets highlighted as light blue (same color when you select objects in Pd)?

since this became a bit harder than I thought... maybe we can do something simpler, like adding a character lke "|" to indicate the box is active - that's easier, cleaner and I might even be able to do it...

BTW, ">" is what [nbx] uses when you're typing in it, while gatom uses "...", is that another inconsistency we should worry about?

**Comment by porres** [comment]
> maybe we can do something simpler, like adding a character lke "|" t

that's really bad as it eats up a character - a light blue highlight color is much more elegant and effective

**Comment by porres** [comment]
> We can also maybe open a new issue to request and discuss this behaviour for atom boxes showing you if it's active

there we go https://github.com/pure-data/pure-data/issues/1013

**Comment by porres** [comment]
> Ok, related to this request, we have:
> - Allow atom number box to type "+" 
> - Remove the activity time of 3 seconds from [nbx], let it keep active until you click out of the object 
> - Keep [nbx] active after hitting enter, so it allows you to type another number.  

Sorry, my last PR had an issue and I made a new one from scratch, this one now fixes this issue, please test: https://github.com/pure-data/pure-data/pull/1015

**Comment by porres** [comment]
> Sorry, my last PR had an issue and I made a new one from scratch, this one now fixes this issue, please test: https://github.com/pure-data/pure-data/pull/1015

new rebooted PR https://github.com/pure-data/pure-data/pull/1689

**Comment by porres** [comment]
Can we close this issue? It's been higly poluted and I am now rebooting it into a new one, which is now a duplicate, see https://github.com/pure-data/pure-data/issues/1747

I will create yet a new PR for this new issue

**Comment by umlaeute** [comment]
we can cleanup "pollution" by hiding the (ir)relevant comments.

let's close the issue once it gets fixed. (although i haven't spend time to see how broken this issue actually is)

**Comment by porres** [comment]
> i haven't spend time to see how broken this issue actually is

not really broken, it's just that now https://github.com/pure-data/pure-data/issues/1747 is a duplicate and a reboot of this one.
